src/server/lib/locale-handler.js

```diff
@@ -0,0 +1,18 @@
+import * as cookie from '../lib/cookie'
+
+/**
+ * Get locale based on body / query param / cookie
+ */
+export default function localeHandler (req, res, cookieOptions) {
+  const { body, query } = req
+  // get locale from body, query or cookie
+  const localeCookieValue = req.cookies[cookieOptions.locale.name] || null
+  const locale = body.locale || query.locale || localeCookieValue || null
+
+  // Save locale in a cookie so that can be used for subsequent requests in signin/signout/callback flow
+  if (locale && (locale !== localeCookieValue)) {
+    cookie.set(res, cookieOptions.locale.name, locale, cookieOptions.locale.options)
+  }
+
+  return locale
```

Comment:
We could instead extend `req.options.locale` to be `mergeDeep({}, defaultTranslations, locales[locale])`, so the internal state `req.options.locale` would only carry around the object containing the actual translations, instead of the locale string and the entire locales object. This would centralize this logic to `localeHandler` and wouldn't need another cookie.

Reply:
We do however need the locale string to keep track of the active locale, no? I'm not quite sure what you mean by this recommended change